File: Binary_Search/Scala/lichtsprung/binary_search.scala
object binary_search {

  def search[T: Ordering](values: Array[T], target: T): Int = {
    val ord = Ordering[T]
    import ord._

    if (values.size < 1) {
      return -1
    }

    val mid = values.size / 2
    val midValue = values(mid)

    if (midValue > target) {
      return search(values.slice(0, mid), target)
    } else if (midValue < target) {
      val p = search(values.slice(mid + 1, values.size), target)
      if (p < 0) return -1 else return mid + 1 + p
    }
    mid
  }

}

File: Binary_Search/Java/lichtsprung/BinarySearch.java
import java.util.Arrays;

public class BinarySearch {
    public static int search(int[] values, int target) {
        if (values.length < 1) {
            return -1;
        }

        int mid = values.length / 2;
        int midValue = values[mid];

        if (midValue > target) {
            return search(Arrays.copyOfRange(values, 0, mid), target);
        } else if (midValue < target) {
            int p = search(Arrays.copyOfRange(values, mid + 1, values.length), target);
            return (p < 0) ? -1 : mid + 1 + p;
        }
        return mid;
    }
}
